### ISSUE TITLE
RetinaNet trains very slowly

### ISSUE THREAD

**Issue by tanzhenyu** [keras-team/keras-cv#1073]
on both TPU and GPU today.

**Comment by tanzhenyu** [comment]
fixed here https://github.com/keras-team/keras-cv/pull/1071

**Comment by bhack** [comment]
Can we clarify this description? It seems that metrics are running slow on TPU but I don't think that they can run TPU right?

I suppose that it is currently slow and not compilable on all the devices right?

**Comment by tanzhenyu** [comment]
> Can we clarify this description? It seems that metrics are running slow on TPU but I don't think that they can run TPU right?
> 
> I suppose that it is currently slow and not compilable on all the devices right?

From what I understand, the issue is COCOMetrics not TPU/GPU compatible, but somehow it was still running as part of the training graph

**Comment by bhack** [comment]
I suppose that if we use `tf.debugging.set_log_device_placement` some of these ops are not placed on these devices or are not XLA compilable.

Other then this I suppose that it is slow also on CPU as I don't think that metrics can run in parallel going forward with training steps right?

So what is the next claim claim?

> In our benchmarks buckets = 10000
achieves strong accuracy, while maintaining tolerable runtime
performance.

In https://arxiv.org/abs/2207.12120

**Comment by tanzhenyu** [comment]
> https://arxiv.org/abs/2207.12120

So in the long run we should just accept there are evaluation that cannot be done on TPU/GPU, and even not TF ops. So I'm proposing a PyMetrics method which allows users to wrap other open source evaluation tools with Keras Metrics. As a first workaround, today we do have PyCOCOCallbacks to help us evaluate our trained models, which is not a public API yet and will be migrated to PyMetrics in the future.

**Comment by bhack** [comment]
> > https://arxiv.org/abs/2207.12120
> 
> So in the long run we should just accept there are evaluation that cannot be done on TPU/GPU, and even not TF ops. So I'm proposing a PyMetrics method which allows users to wrap other open source evaluation tools with Keras Metrics. As a first workaround, today we do have PyCOCOCallbacks to help us evaluate our trained models, which is not a public API yet and will be migrated to PyMetrics in the future.

Ok but I don't think that this design will solve the speed issue. Do you think that we are going to accumulate c++/CUDA custom ops on metrics?

**Comment by tanzhenyu** [comment]
> > > https://arxiv.org/abs/2207.12120
> > 
> > 
> > So in the long run we should just accept there are evaluation that cannot be done on TPU/GPU, and even not TF ops. So I'm proposing a PyMetrics method which allows users to wrap other open source evaluation tools with Keras Metrics. As a first workaround, today we do have PyCOCOCallbacks to help us evaluate our trained models, which is not a public API yet and will be migrated to PyMetrics in the future.
> 
> Ok but I don't think that this design will solve the speed issue. Do you think that we are going to accumulate c++/CUDA custom ops on metrics?

I haven't noticed any obvious speed issue for the callbacks yet. I'm yet to see how other GLUE like metrics are being implemented to decide whether we want to support pure python metrics.

**Comment by bhack** [comment]
> on both TPU and GPU today, it took an hour for the 1st epoch.

So do you mean that the callback is going to solve this?

**Comment by tanzhenyu** [comment]
> > on both TPU and GPU today, it took an hour for the 1st epoch.
> 
> So do you mean that the callback is going to solve this?

As a temporary solution, yes
